aiidalab_widgets_base/databases.py

```diff
@@ -204,179 +204,234 @@ def _update_structure(self, change: dict) -> None:
-            # If no computers remained - remove the domain.
-            if len(database[domain].keys() - {"default"}) == 0:
+                        # remove code
+                        del computer_value["codes"][code]
+
+                if len(computer_value["codes"]) == 0:
+                    # remove computer since no codes defined in this computer source
+                    del domain_value[computer]
+                    if domain_value.get("default") == computer:
+                        # also remove default computer from domain
+                        del domain_value["default"]
+
+            if len(domain_value) == 0:
+                # remove domain since no computers with required codes defined in this domain source
                 del database[domain]
-            # Making sure the 'default' key still points to an existing computer.
-            elif database[domain]["default"] not in database[domain]:
-                database[domain]["default"] = sorted(
-                    database[domain].keys() - {"default"}
-                )[0]
-        return database
+                continue
 
-    def update(self, _=None):
-        database = requests.get(
-            "https://aiidateam.github.io/aiida-code-registry/database_v2_1.json"
-        ).json()
-        self.database = (
-            self.clean_up_database(database, self.default_calc_job_plugin)
-            if self.default_calc_job_plugin
-            else database
-        )
+            if domain_value["default"] not in domain_value:
+                # make sure default computer is still points to existing computer
+                domain_value["default"] = sorted(domain_value.keys() - {"default"})[0]
 
-    @tl.observe("database")
-    def _observer_database_change(self, _=None):
-        self.inp_domain.options = self.database.keys()
-        self._reset()
+        return database
 
-    def _domain_changed(self, _=None):
+    def _domain_changed(self, change=None):
+        """callback when new domain selected"""
+        with self.hold_trait_notifications():  # To prevent multiple calls to callbacks
+            if change["new"] is None:
+                self.computer_selector.options = ()
+                self.computer_selector.value = None
+                self.code_selector.options = ()
+                self.code_selector.value = None
+                return
+            else:
+                selected_domain = self.domain_selector.value
+
+            with self.hold_trait_notifications():
+                try:
+                    self.computer_selector.options = tuple(
+                        key
+                        for key in self.database[selected_domain].keys()
+                        if key != "default"
+                    )
+                    self.computer_selector.value = self.database[selected_domain][
+                        "default"
+                    ]
+                except KeyError:
+                    raise
```

Comment:
Nice catch, it was the left over code block from old implementation. Since I use dict get with default empty dict returned as default, those try..except clause are not hit anymore.